### PR TITLE
add dataset download

### PR DESCRIPTION
**Pull request by liuyhwangyh**
1. add dataset_file_download dataset_snapshot_download
2. fix model_file_download bug of file in sub directory
3. add datasets extra requirement.
4. framework dependencies refine
5. add triton and lmdeploy to image.

### CODE REVIEW

File: modelscope/cli/download.py [hunk 1]
```diff
@@ -3,6 +3,8 @@
 from argparse import ArgumentParser
 
 from modelscope.cli.base import CLICommand
+from modelscope.hub.dataset_download import (dataset_file_download,
```
Comment (yingdachen) [inline modelscope/cli/download.py:6]: dataset_file_download 放到file_download里面（和model_file_download放一起）？

File: modelscope/cli/download.py [hunk 2]
```diff
@@ -24,11 +26,17 @@ def define_args(parsers: ArgumentParser):
...
         """
         parser: ArgumentParser = parsers.add_parser(DownloadCMD.name)
-        parser.add_argument(
+        group = parser.add_mutually_exclusive_group(required=True)
+        group.add_argument(
             '--model',
             type=str,
-            required=True,
-            help='The model id to be downloaded.')
+            help='The model id to be downloaded, model or dataset must provide.'
```
Comment (yingdachen) [inline modelscope/cli/download.py:33]: The id of the model to be downloaded. For download, the id of either a model or dataset must be provided.

File: modelscope/cli/download.py [hunk 3]
```diff
@@ -24,11 +26,17 @@ def define_args(parsers: ArgumentParser):
...
+        )
+        group.add_argument(
+            '--dataset',
+            type=str,
+            help=
+            'The dataset id to be downloaded, model or dataset must provide.')
```
Comment (yingdachen) [inline modelscope/cli/download.py:39]: The id of the dataset to be downloaded. For download, the id of either a model or dataset must be provided.

File: modelscope/cli/download.py [hunk 4]
```diff
@@ -69,27 +77,55 @@ def define_args(parsers: ArgumentParser):
...
+            else:  # download repo
+                snapshot_download(
+                    self.args.model,
+                    revision=self.args.revision,
+                    cache_dir=self.args.cache_dir,
+                    local_dir=self.args.local_dir,
+                    allow_file_pattern=self.args.include,
+                    ignore_file_pattern=self.args.exclude,
+                )
+        else:
```
Comment (yingdachen) [inline modelscope/cli/download.py:106]: else if self.args.dataset is not None ?

File: modelscope/cli/download.py [hunk 5]
```diff
@@ -69,27 +77,55 @@ def define_args(parsers: ArgumentParser):
...
+                    local_dir=self.args.local_dir,
+                    allow_file_pattern=self.args.files,
+                )
+            else:  # download repo
+                snapshot_download(
+                    self.args.model,
+                    revision=self.args.revision,
+                    cache_dir=self.args.cache_dir,
+                    local_dir=self.args.local_dir,
+                    allow_file_pattern=self.args.include,
```
Comment (yingdachen) [inline modelscope/cli/download.py:103]: 这里是下载整个repo吗？那为什么还会有file pattern？

File: modelscope/hub/dataset_download.py [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+                        MODELSCOPE_PARALLEL_DOWNLOAD_THRESHOLD_MB)
+from .file_download import (create_temporary_directory_and_cache,
+                            http_get_model_file, parallel_download)
+from .utils.utils import (file_integrity_validation,
+                          model_id_to_group_owner_name)
+
+logger = get_logger()
+
+
+def dataset_file_download(
```
Comment (yingdachen) [inline modelscope/hub/dataset_download.py:24]: 这里的逻辑是不是大部分和model_file_download是一致的？建议复用一下代码？

dataset_snapshot_download同理

File: modelscope/hub/file_download.py [hunk 7]
```diff
@@ -269,6 +270,7 @@ def parallel_download(
     PART_SIZE = 160 * 1024 * 1024  # every part is 160M
     tasks = []
     file_path = os.path.join(local_dir, file_name)
+    os.makedirs(os.path.dirname(file_path), exist_ok=True)
```
Comment (yingdachen) [inline modelscope/hub/file_download.py:273]: 这里是修复下载子目录文件的bug？

File: modelscope/utils/file_utils.py [hunk 8]
```diff
@@ -53,11 +53,35 @@ def get_model_cache_root() -> str:
     """Get model cache root path.
 
     Returns:
-        str: the modelscope cache root.
+        str: the model cache root.
```
Comment (yingdachen) [inline modelscope/utils/file_utils.py:56]: modelscope model cache root

File: requirements/cv.txt [hunk 9]
```diff
@@ -44,10 +44,11 @@ opencv-python
 paint_ldm
 pandas
 panopticapi
+Pillow>=6.2.0
 plyfile>=0.7.4
 psutil
 pyclipper
-PyMCubes
+PyMCubes<=0.1.4
```
Comment (yingdachen) [inline requirements/cv.txt:51]: 这里为什么要限定最高版本？

File: requirements/cv.txt [hunk 10]
```diff
@@ -44,10 +44,11 @@ opencv-python
 paint_ldm
 pandas
 panopticapi
+Pillow>=6.2.0
```
Comment (yingdachen) [inline requirements/cv.txt:47]: cv应该基于framework吧？framework理由的东西就不需要了？

File: modelscope/__init__.py [hunk 11]
```diff
@@ -60,6 +62,8 @@
         ],
         'hub.api': ['HubApi'],
         'hub.snapshot_download': ['snapshot_download'],
+        'hub.dataset_download':
+        ['dataset_snapshot_download', 'dataset_file_download'],
```
Comment (wenmengzhou) [inline modelscope/__init__.py:66]: dataset_snapshot_download  -> dataset_download   may be better?

File: modelscope/hub/dataset_download.py [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+        temp_file = os.path.join(temporary_cache_dir,
+                                 file_meta_to_download['Name'])
+        if FILE_HASH in file_meta_to_download:
+            file_integrity_validation(temp_file,
+                                      file_meta_to_download[FILE_HASH])
+        # put file into to cache
+        return cache.put_file(file_meta_to_download, temp_file)
+
+
+def dataset_snapshot_download(
```
Comment (wenmengzhou) [inline modelscope/hub/dataset_download.py:169]: can we reuse the original implementation of snapshot_download?

File: tests/hub/test_download_dataset_file.py [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+                                             dataset_snapshot_download)
+
+
+class DownloadDatasetTest(unittest.TestCase):
+
+    def setUp(self):
+        pass
+
+    def test_dataset_file_download(self):
+        dataset_id = 'citest/test_dataset_download'
```
Comment (wenmengzhou) [inline tests/hub/test_download_dataset_file.py:18]: also test wildcard characters

File: modelscope/hub/dataset_download.py (hunk 12 above)
Comment (yingdachen) [reply modelscope/hub/dataset_download.py:169]: > can we reuse the original implementation of snapshot_download?

originally i had the same question, but our snapshot_download takes a named parameter "model_id", so it is hard to reuse the same function (in comparison, hf's parameter is the more generic repo_id, which can refer to both model and dataset). i think the best we can do is to refactor the major body of the implementation, and have both the original snapshot_download and dataset_snapshot_download sharing this underlying implementation...

File: modelscope/__init__.py (hunk 11 above)
Comment (liuyhwangyh) [reply modelscope/__init__.py:66]: dataset_file_download and dataset_snapshot_download are a pair .

File: requirements/cv.txt (hunk 9 above)
Comment (liuyhwangyh) [reply requirements/cv.txt:51]: cv安装时最新版有兼容性问题。

File: requirements/cv.txt (hunk 10 above)
Comment (liuyhwangyh) [reply requirements/cv.txt:47]: Pillow在cv领域用，放在framework里感觉不合适。

File: tests/hub/test_download_dataset_file.py (hunk 13 above)
Comment (liuyhwangyh) [reply tests/hub/test_download_dataset_file.py:18]: It has been added in dataset_snapshot_download,  dataset_file_download only supports exact path.

File: modelscope/hub/dataset_download.py (hunk 12 above)
Comment (liuyhwangyh) [reply modelscope/hub/dataset_download.py:169]: 骨架差不多，model和dataset用了两套接口，目前看单独放在一个文件便于维护，底层下载部分已经尽可能复用了。

File: modelscope/utils/file_utils.py (hunk 8 above)
Comment (liuyhwangyh) [reply modelscope/utils/file_utils.py]: updated

File: modelscope/hub/file_download.py (hunk 7 above)
Comment (liuyhwangyh) [reply modelscope/hub/file_download.py:273]: Yes

File: modelscope/hub/dataset_download.py (hunk 6 above)
Comment (liuyhwangyh) [reply modelscope/hub/dataset_download.py:24]: 骨架差不多，model和dataset用了两套接口，目前看单独放在一个文件便于维护，底层下载部分已经尽可能复用了。

File: modelscope/cli/download.py (hunk 5 above)
Comment (liuyhwangyh) [reply modelscope/cli/download.py:103]: 类似model，下载整个repo也支持过滤某些文件或者只下载指定文件。
modelscope download --model 'AI-ModelScope/gpt2' --include 'onnx/*' '*.tflite'

File: modelscope/cli/download.py (hunk 4 above)
Comment (liuyhwangyh) [reply modelscope/cli/download.py]: 改了，加上条件更清晰，group已经设置成必选了，这两个参数必须输入一个，效果都一样。

File: modelscope/cli/download.py (hunk 3 above)
Comment (liuyhwangyh) [reply modelscope/cli/download.py]: updated

File: modelscope/cli/download.py (hunk 2 above)
Comment (liuyhwangyh) [reply modelscope/cli/download.py]: updated

File: modelscope/cli/download.py (hunk 1 above)
Comment (liuyhwangyh) [reply modelscope/cli/download.py:6]: 看起来逻辑一样，但model和dataset用了两套接口，目前看单独放在一个文件便于维护，底层下载部分已经尽可能复用了。

File: modelscope/hub/dataset_download.py (hunk 12 above)
Comment (wenmengzhou) [reply modelscope/hub/dataset_download.py:169]: agreed， should we use repo_id and  snapshot_download   for both dataset repo and model repo by switching the service endpoint for dataset hub and modelhub?

File: requirements/cv.txt (hunk 10 above)
Comment (yingdachen) [reply requirements/cv.txt:48]: ah, i see you have removed it from framework.txt... OK then

**Review by yingdachen — APPROVED** [review]
LGTM

File: modelscope/cli/download.py (hunk 1 above)
Comment (liuyhwangyh) [reply modelscope/cli/download.py:6]: 合并了

File: modelscope/hub/dataset_download.py (hunk 6 above)
Comment (liuyhwangyh) [reply modelscope/hub/dataset_download.py]: 合并了

File: modelscope/hub/dataset_download.py (hunk 12 above)
Comment (liuyhwangyh) [reply modelscope/hub/dataset_download.py]: 更新了

**Pull request merged**